staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,370 @@
+	return h.finishedSync.Load()
+}
+
+func (h *peerProxyHandler) WaitForCacheSync(stopCh <-chan struct{}) error {
+
+	ok := cache.WaitForNamedCacheSync("unknown-version-proxy", stopCh, h.storageversionInformer.HasSynced, h.storageversionManager.Completed)
+	if !ok {
+		return fmt.Errorf("error while waiting for initial cache sync")
+	}
+	klog.V(3).Infof("setting finishedSync to true")
+	h.finishedSync.Store(true)
+	return nil
+}
+
+// Handle will fetch the apiservers that can serve the request and either serve it locally
+// or route it to a peer
+func (h *peerProxyHandler) Handle(handler http.Handler, serverId string, s runtime.NegotiatedSerializer, reconciler reconcilers.PeerEndpointLeaseReconciler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+
+		if !ok {
+			responsewriters.InternalError(w, r, errors.New("no RequestInfo found in the context"))
+			return
+		}
+
+		// Allow non-resource requests
+		if !requestInfo.IsResourceRequest {
+			klog.V(3).Infof("Not a resource request skipping proxying")
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		// Request has already been proxied once, it must be served locally
+		if r.Header.Get(PeerProxiedHeader) == "true" {
+			klog.V(3).Infof("Already rerouted once, skipping proxying to peer")
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		// StorageVersion Informers and/or StorageVersionManager is not synced yet, pass request to next handler
+		// This will happen for self requests from the kube-apiserver because we have a poststarthook
+		// to ensure that external requests are not served until the StorageVersion Informer and
+		// StorageVersionManager has synced
+		if !h.HasFinishedSync() {
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		gvr := schema.GroupVersionResource{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion, Resource: requestInfo.Resource}
+		if requestInfo.APIGroup == "" {
+			gvr.Group = "core"
+		}
+
+		// find servers that are capable of serving this request
+		serviceableByResp, err := h.findServiceableByServers(gvr, serverId, reconciler)
+		if err != nil {
+			// this means that resource is an aggregated API or a CR since it wasn't found in SV informer cache, pass as it is
+			klog.Infof(fmt.Sprintf("no StorageVersion/APIServerID found for the GVR: %v skipping proxying", gvr))
+			handler.ServeHTTP(w, r)
+			return
+		}
+		// found the gvr locally, pass request to the next handler in local apiserver
+		if serviceableByResp.locallyServiceable {
+			klog.V(4).Infof("resource can be served locally, skipping proxying")
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		gv := schema.GroupVersion{Group: gvr.Group, Version: gvr.Version}
+
+		// if no apiservers were found that could serve the request, serve 404
+		if len(serviceableByResp.peerEndpoints) == 0 {
+			klog.Errorf(fmt.Sprintf("GVR %v is not served by anything in this cluster", gvr))
```

Comment:
Request for unknown resources should hit [here](https://github.com/kubernetes/kubernetes/blob/7f9a8b6feb335629c9eb33828fc28f807f4ed22b/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go#L203) and so eventually [here](https://github.com/kubernetes/kubernetes/blob/7f9a8b6feb335629c9eb33828fc28f807f4ed22b/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go#L151-L156)

for that scenario, we pass the request to next handler in the chain since it can be for an aggregated API/CR